### DISCUSSION TITLE
Independent versioning

### DISCUSSION THREAD

**Discussion by heinrichb-outsurance-ie** [leonardochaia/dotnet-affected#92] [Q&A]
Hi, 
Was wondering what all the peeps are using for versioning.
Just 1 version per repository (GitVersion) or are there a solution to independently version certain packages/projects.

Thanks in advance

**Comment by leonardochaia** [comment]
Hi there. We  do both..

Monorepos work smooth with trunk based development. So we have our main branch, and we branch off there to create feature branches with specific features. These then get merged back to main using pull requests. That's how we introduce new changes.

For releasing, we version all components with the same version when we release, by creating a git tag over the main branch. This triggers a pipeline which builds all docker images and tags then with the git tag
`main` is like the develop branch where everything is happening. When we wanna cut a release, we git tag main with the version name.
When we need to fix a bug of a previous release, we fix it on main first, then we branch off the buggy release tag, cherry pick the fix commit (if applicable), then create the new minor tag pointing to the cherry picked commit of the buggy release branch. Then we can just delete the branch and keep the tag.

For development, when a new commit is merged to main, a pipeline is triggered which runs dotnet-affected to see what was actually affected. In that case, we create docker images with the current build id as a tag only for the affected services. Then argocd-image-updater takes care of commiting the new version to our dev environment.

We also have a nightly build daily which builds all docker images with the build id tag so that every day we start off with a fresh set of images which have the same tag, instead of a messy multiple tags environment.

Happy to answer further questions.

Kind regards,
Leo.
